---
title: "LDA tutorial"
output: html_document
date: "2022-09-14"
---
This Linear Discriminant Analysis tutorial working on RefSE tool on mothur by Rstudio platform.
1- Library
```{r}
library(tidyverse)
library(readxl)
library(glue)
library(qiime2R)
library(ggtext)
```
2- LDA calculator using LefSe-mothur
2.1.Preparing file for mothur
```{r}
metadata <- read_excel("sample-metadata.xlsx") %>% select (`sample-id`, `Type`)
SVs<-read_qza("table.qza")$data
taxonomy<-read_qza("taxonomy.qza")$data %>% parse_taxonomy()
taxasums<-summarize_taxa(SVs, taxonomy)$Genus
shared_file <- taxasums
write.csv(shared_file, "shared_file.csv") 
#After that add "group" in title
#Transpose data
#Add label 0.03 column (1st)
#Add ASVs column(number of ASV) ~ 191 in this case
#change name classifier to ASVi (i=1>191)
shared_file <- read_excel("shared_file.xlsx")
shared_design <- inner_join(shared_file, metadata, by=c("Group" = "sample-id"))
write_tsv(shared_design,"shared_design.tsv")
```
2.2.LDA-calculator-automatic-function-by-LefSe-Mothur
```{r}
dir.create("processed_data")
run_lefse <- function (x,y,tag){
x_y <- shared_design %>%
  filter(Type == "x" | Type == "y")
x_y %>%
 select (-`Type`) %>%
 write_tsv(glue("processed_data/{tag}.shared"))
x_y %>%
  select(`Group`, `Type`) %>%
  write_tsv(glue("processed_data/{tag}.design"))
command <- glue('i:/Metagenomic-tools/mothur/mothur "#lefse(shared={tag}.shared, design={tag}.design, inputdir=processed_data)"')
system(command)
return(glue("processed_data/{tag}.0.03.lefse_summary"))
}
```
3- Modify taxasums
```{r}
SVs<-read_qza("table.qza")$data
taxonomy<-read_qza("taxonomy.qza")$data %>% parse_taxonomy()
taxasums<-summarize_taxa(SVs, taxonomy)$Genus
write.csv(taxasums,"taxasums.csv")
```
#Using excel to: Add ASVs column in taxasums, save as tsv, add Group in name of bacterial column
```{r}
taxonomy <- read_excel("taxasums.xlsx") %>%
  select("ASVs", "Group") %>%
  rename_all(tolower) %>%
  separate(group,
           into=c("kingdom", "phylum", "class", "order", "family", "genus"),
           sep=";") %>%
        mutate(genus = str_replace(string=genus,   #add ** around
                             pattern="(.*)", 
                             replacement="*\\1*"),
               taxon = glue("{genus}<br>({asvs})")) %>%
  select(asvs, taxon)
```
4.LDA-plot
```{r}
LDA<-read_tsv("processed_data/MA_MM.0.03.lefse_summary") %>% 
drop_na(LDA) %>%
  filter (LDA > 3) %>%
inner_join(., taxonomy, by=c("OTU" = "asvs")) %>%
mutate (LDA = if_else(Class =="MM", -1*LDA, LDA), taxon = fct_reorder(taxon,LDA))

ggplot(LDA, aes(x=LDA, y=taxon, fill=Class)) + geom_col() + labs (y=NULL, x="LDA Score (log 10)") + scale_x_continuous(limits = c(-6,6), breaks = seq(-6, 6, by=2)) + theme_classic()
```
